OneCommunityGlobal/HGNRest
# Shengwei_eliminate_sentry_error_and_update_sentry_log_plugin

# Description
**Issue:** 
- **Lack of log separation and isolation**. Currently, all error logs from local, dev, and prod are sent to Sentry, leading to the "Noisy" Logging Problem. This makes it challenging to monitor production issues and accurately identify their origin.
- The **outdated Sentry library** version is not capturing enough information, such as request info and sufficient stack traces.
- While most errors in the backend application are handled, they are **not properly logged to Sentry**, resulting in difficulties tracking and monitoring errors in the production environment.


## Related PRS (if any):

## Main changes explained:
- Introduce @sentry/integrations and update Sentry logging configuration.
- Update file ../startup/logger.js to include log separation.
- Update @sentry/node from  `5.17.0 ` to `7.56.0`
- Add Sentry event middlewares in `src/server.js`


## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally
3. Create a new route and API for log testing
   - ```
     const Logger = require('../startup/logger');
     ...
     try {
        aFunctionThatMightFail();
      } catch (e) {
        Logger .logException(err, transactionName, extraData);
      }
     ```
4. To test for local and development environment logging, you need to configure NODE_ENV by adding **NODE_ENV=local** or **NODE_ENV=development** in the `.env` file. 
5. Logs do not to Sentry in `local` environment. Logs in any other environments send to Sentry and searchable by tag 'environment'
<img width="939" alt="image" src="https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/27e42852-7150-4459-ad75-e4c9a41113c7">

## Screenshots or videos of changes:
### Comparison of old and new Sentry log
**New Data Captured:**
- OS version
- Runtime
- Transaction name
- URL
- Server status
- Customizable additional data
- Client IP (Currently turnoff)
![sentry_comparison](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/dbdfb5b6-90c2-4fc5-bb27-e7b8a19dfeee)

### Configurable 'Extra Data' 
<img width="780" alt="image" src="https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/ff8e86b1-6f2a-4dc8-a8d2-6e9bb8275095">
<img width="587" alt="image" src="https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/fdd56ea6-bd1a-414a-a14e-2963f0038ebb">

## Note:
Include the information the reviewers need to know.
